pkg/alertmanager/alertstore/bucketclient/protobuf_store.go

```diff
@@ -0,0 +1,95 @@
+package bucketclient
+
+import (
+	"bytes"
+	"context"
+	"io/ioutil"
+
+	"github.com/go-kit/kit/log"
+	"github.com/gogo/protobuf/proto"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"github.com/thanos-io/thanos/pkg/runutil"
+
+	"github.com/cortexproject/cortex/pkg/storage/bucket"
+)
+
+// ProtobufStore is used to read/write protobuf messages to object storage backend. It is implemented
+// using the Thanos objstore.Bucket interface
+type ProtobufStore struct {
+	bucket      objstore.Bucket
+	cfgProvider bucket.TenantConfigProvider
+	logger      log.Logger
+}
+
+func NewProtobufStore(prefix string, bkt objstore.Bucket, cfgProvider bucket.TenantConfigProvider, logger log.Logger) *ProtobufStore {
+	return &ProtobufStore{
+		bucket:      bucket.NewPrefixedBucketClient(bkt, prefix),
+		cfgProvider: cfgProvider,
+		logger:      logger,
+	}
+}
+
+// ListAllUsers returns all users with an entry in the store.
+func (s *ProtobufStore) ListAllUsers(ctx context.Context) ([]string, error) {
+	var userIDs []string
+
+	err := s.bucket.Iter(ctx, "", func(key string) error {
+		userIDs = append(userIDs, key)
+		return nil
+	})
+
+	return userIDs, err
+}
+
+// Set writes the content for an entry in the store for a particular user.
+func (s *ProtobufStore) Set(ctx context.Context, userID string, msg proto.Message) error {
```

Comment:
What do you think about calling it `Upload()` to keep it specular to the underlying `objstore.Bucket` function name?